cardano-testnet/test/cardano-testnet-test/Cardano/Testnet/Test/Api/TxReferenceInputDatum.hs

```diff
@@ -0,0 +1,243 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE NumericUnderscores #-}
+{-# LANGUAGE OverloadedLists #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Cardano.Testnet.Test.Api.TxReferenceInputDatum
+  ( hprop_tx_refin_datum
+  )
+where
+
+import           Cardano.Api
+import qualified Cardano.Api.Ledger as L
+import qualified Cardano.Api.Network as Net
+import qualified Cardano.Api.Network as Net.Tx
+import           Cardano.Api.Shelley
+
+import           Cardano.Testnet
+
+import           Prelude
+
+import           Control.Monad
+import           Data.Default.Class
+import qualified Data.Map.Strict as M
+import           Data.Proxy
+import           Data.Set (Set)
+import           GHC.Exts (IsList (..))
+import           GHC.Stack
+import           Lens.Micro
+
+import           Testnet.Components.Query
+import           Testnet.Property.Util (integrationRetryWorkspace)
+import           Testnet.Types
+
+import           Hedgehog
+import qualified Hedgehog as H
+import qualified Hedgehog.Extras.Test.Base as H
+import qualified Hedgehog.Extras.Test.TestWatchdog as H
+
+
+-- | Test providing actual datum to reference inputs containing datum hashes
```

Comment:
I don't understand what this means. Can you elaborate on what this test is testing and how this differs from "normal" datum usage.

Reply:
Added more comments.